fork_choice/confirmation-rule.md

```diff
@@ -0,0 +1,639 @@
+    assert current_epoch in [block_epoch, block_epoch + 1]
+
+    checkpoint_root = get_checkpoint_block(store, block_root, block_epoch)
+    checkpoint = Checkpoint(root=checkpoint_root, epoch=block_epoch)
+
+    if checkpoint not in store.checkpoint_states:
+        return Gwei(0)
+
+    checkpoint_state = store.checkpoint_states[checkpoint]
+
+    leaf_roots = get_leaf_block_roots(store, block_root)
+
+    active_checkpoint_indices = get_active_validator_indices(checkpoint_state, block_epoch)
+    participating_indices = set().union(*[
+        get_epoch_participating_indices(
+            store.block_states[root], 
+            active_checkpoint_indices, 
+            block_epoch == current_epoch
+        )
+        for root in leaf_roots
+    ])
+
+    return get_total_balance(checkpoint_state, participating_indices)
+```
+
+
+#### `is_ffg_confirmed`
+
+```python
+def is_ffg_confirmed(store: Store, block_root: Root) -> bool:
+    """
+    Returns whether the `block_root`'s checkpoint will be justified by the end of this epoch.
+    """
+    current_epoch = get_current_store_epoch(store)
+
+    block = store.blocks[block_root]
+    block_epoch = compute_epoch_at_slot(block.slot)
+
+    # This function is only applicable to current and previous epoch blocks
+    assert current_epoch in [block_epoch, block_epoch + 1]
+
+    total_active_balance = int(get_total_active_balance_for_block_root(store, block_root))
+
+    ffg_support_for_checkpoint = int(get_ffg_support(store, block_root))
+
+    if block_epoch == current_epoch:
+        remaining_ffg_weight = int(get_remaining_weight_in_current_epoch(store, block_root))
+    else:
+        remaining_ffg_weight = 0
+
+    max_adversarial_ffg_support_for_checkpoint = int(
+        min(
+            (total_active_balance * CONFIRMATION_BYZANTINE_THRESHOLD - 1) // 100 + 1,
+            CONFIRMATION_SLASHING_THRESHOLD,
+            ffg_support_for_checkpoint
+        )
+    )
+
+    # Returns whether the following condition is true using only integer arithmetic
+    # 2 / 3 * total_active_balance <= (
+    #     ffg_support_for_checkpoint - max_adversarial_ffg_support_for_checkpoint +
+    #     (1 - CONFIRMATION_BYZANTINE_THRESHOLD / 100) * remaining_ffg_weight
+    # )
+
+    return (
+        200 * total_active_balance <=
+        ffg_support_for_checkpoint * 300 + (300 - 3 * CONFIRMATION_BYZANTINE_THRESHOLD) *
+        remaining_ffg_weight - max_adversarial_ffg_support_for_checkpoint * 300
+    )
+```
+
+### `is_confirmed`
+
+```python
+def is_confirmed(store: Store, block_root: Root) -> bool:
+    current_epoch = get_current_store_epoch(store)
+
+    block = store.blocks[block_root]
+    block_state = store.block_states[block_root]
+    block_justified_checkpoint_epoch = block_state.current_justified_checkpoint.epoch
+    block_epoch = compute_epoch_at_slot(block.slot)
+
+    # This function is only applicable to current and previous epoch blocks
```

Comment:
Is there a way to get confirmation for a block that belongs to an epoch that is older than the previous one?

Reply:
Resolving this as we can discuss about this in [this other thread](https://github.com/ethereum/consensus-specs/pull/3339#discussion_r1280108401).